cmd/restic/cmd_rewrite.go

```diff
@@ -0,0 +1,296 @@
+	f.BoolVarP(&rewriteOptions.Inplace, "inplace", "", false, "replace existing snapshots")
+	f.BoolVarP(&rewriteOptions.DryRun, "dry-run", "n", false, "do not do anything, just print what would be done")
+
+	// Excludes
+	f.StringArrayVarP(&rewriteOptions.Excludes, "exclude", "e", nil, "exclude a `pattern` (can be specified multiple times)")
+	f.StringArrayVar(&rewriteOptions.InsensitiveExcludes, "iexclude", nil, "same as --exclude `pattern` but ignores the casing of filenames")
+	f.StringArrayVar(&rewriteOptions.ExcludeFiles, "exclude-file", nil, "read exclude patterns from a `file` (can be specified multiple times)")
+}
+
+type saveTreeFunction = func(*restic.Tree) (restic.ID, error)
+
+func filterNode(ctx context.Context, repo restic.Repository, nodepath string, nodeID restic.ID, checkExclude RejectByNameFunc, saveTreeFunc saveTreeFunction) (newNodeID restic.ID, err error) {
+	curTree, err := repo.LoadTree(ctx, nodeID)
+	if err != nil {
+		return nodeID, err
+	}
+
+	debug.Log("filterNode: %s, nodeId: %s\n", nodepath, nodeID.Str())
+
+	changed := false
+	newTree := restic.NewTree()
+	for _, node := range curTree.Nodes {
+		path := path.Join(nodepath, node.Name)
+		if !checkExclude(path) {
+			if node.Subtree == nil {
+				newTree.Insert(node)
+				continue
+			}
+			newNode := node
+			newID, err := filterNode(ctx, repo, path, *node.Subtree, checkExclude, saveTreeFunc)
+			if err != nil {
+				return nodeID, err
+			}
+			if newID == *node.Subtree {
+				newTree.Insert(node)
+			} else {
+				changed = true
+				newNode.Subtree = new(restic.ID)
+				*newNode.Subtree = newID
+				newTree.Insert(newNode)
+			}
+		} else {
+			Verbosef("excluding %s\n", path)
+			changed = true
+		}
+	}
+
+	if changed {
+		// Save new tree
+		newTreeID, err := saveTreeFunc(newTree)
+		debug.Log("filterNode: save new tree for %s as %v\n", nodepath, newTreeID)
+		return newTreeID, err
+	}
+
+	return nodeID, nil
+}
+
+func collectRejectFuncsForRewrite(opts RewriteOptions) (fs []RejectByNameFunc, err error) {
+	//TODO: merge with cmd_backup
+
+	// add patterns from file
+	if len(opts.ExcludeFiles) > 0 {
+		excludes, err := readExcludePatternsFromFiles(opts.ExcludeFiles)
+		if err != nil {
+			return nil, err
+		}
+		opts.Excludes = append(opts.Excludes, excludes...)
+	}
+
+	if len(opts.InsensitiveExcludes) > 0 {
+		fs = append(fs, rejectByInsensitivePattern(opts.InsensitiveExcludes))
+	}
+
+	if len(opts.Excludes) > 0 {
+		fs = append(fs, rejectByPattern(opts.Excludes))
+	}
+
+	return fs, nil
+}
+
+func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *restic.Snapshot, opts RewriteOptions, gopts GlobalOptions) (bool, error) {
+	if sn.Tree == nil {
+		return false, errors.Errorf("snapshot %v has nil tree", sn.ID().Str())
+	}
+
+	var saveTreeFunc saveTreeFunction
+	if !opts.DryRun {
+		saveTreeFunc = func(tree *restic.Tree) (restic.ID, error) {
+			return repo.SaveTree(ctx, tree)
```

Comment:
You could also use the TreeSaver from `internal/archiver` to make saving parallel. Can be however postponed to a later optimization.

Reply:
That would also require the BlobSaver. It's probably better to integrate some form of asynchronous blob submission directly into the repository and then use that.